src/backend/u64/field.rs

```diff
@@ -351,6 +353,113 @@ impl<'a> Half for &'a FieldElement {
+            } else {
+                expon = expon - FieldElement::one();
+            res = res * base;
+
+            expon = expon.half();
+            base = &base * &base;
+            }
+        }
+
+        res
+    }
+}
+
+
+impl<'a> ModSqrt for &'a FieldElement {
+    type Output = Option<FieldElement>;
+    /// Performs the op: `sqrt(a) (mod l)`.
+    /// 
+    /// Tonelli-Shanks prime modular square root
+    /// algorithm implementation for `FieldElement`.
+    /// 
+    /// Conditionally selects and returns the positive or the 
+    /// negative result of the `mod_sqrt` by analyzing the 
+    /// `Choice` sent as input:
+    /// 
+    /// For Choice(0) -> Negative result. 
+    /// For Choice(1) -> Positive result.
+    /// 
+    /// 
+    /// Daniel Shanks. Five Number Theoretic Algorithms. 
+    /// Proceedings of the Second Manitoba Conference on 
+    /// Numerical Mathematics. Pp. 51–70. 1973.
+    /// 
+    /// This algorithm was translated from the python impl
+    /// found in: 
+    /// https://codereview.stackexchange.com/questions/43210/tonelli-shanks-algorithm-implementation-of-prime-modular-square-root
+    fn mod_sqrt(self, sign: Choice) -> Option<FieldElement> {
+        // If the input is `0` the sqrt is directly 0.
+        if bool::from(self.ct_eq(&FieldElement::zero())) 
+            {return Some(FieldElement::zero());}
+
+        // Check if exists a solution insine the finite
+        // field generated by `FIELD_L`. 
+        if !bool::from(self.legendre_symbol()) {return None;}
+
+        // Factor p-1 on the form q * 2^s (with Q odd). 
+        let mut q = FieldElement::minus_one();
+        let mut s = FieldElement::zero(); 
+        while q.is_even() {
+            s = s + FieldElement::one();
+            q = q.half();
+        }
+
+        // Select a z which is a quadratic non resudue modulo p.
+        // We pre-computed it so we know that 6 isn't QR. 
+        let z = FieldElement::from(&6u8); 
+        let mut c = z.pow(&q);
+
+        // Search for a solution. 
+        let mut x = self.pow(&(q + FieldElement::one()).inner_half());
+        let mut t = self.pow(&q);
+        let mut m = s.clone();
+
+        let one = FieldElement::one();
+        let two = FieldElement::from(&2u8);
+
+        while t != FieldElement::one() {
+            // Find the lowest i such that t^(2^i) = 1. 
+            let mut i = FieldElement::zero();
+            let mut e = FieldElement::from(&2u8);
+            let mut b;
+            while i < m {
+                i = i + one;
+                if bool::from(t.pow(&e).ct_eq(&one)) {break;}
+                e = e * two;
+            }
+
+            // Update values for next iter
+            b = c.pow(&two.pow(&(m - i - one)));
+            x = x * b;
+            t = t * b.square();
+            c = b.square();
+            m = i;
+        };
+
+        Some(FieldElement::conditional_select(&x, &(constants::FIELD_L - x), sign))
+    }
+}
+
```

Comment:
Before the completion of this, have the two algorithms been benchmarked? If so, may I ask where?

Reply:
You're right, they should be benchmarked. So there you have them! hehe

`Pow` and `ModSqrt` benchmarked in ea5c535

Point generation benchmarked in 59770b7